hypothesis-python/src/hypothesis/internal/conjecture/providers.py

```diff
@@ -205,15 +208,54 @@
     "bytes": SortedSet(),
     "string": SortedSet(),
 }
-# modules that we've already seen and processed for local constants.
-_local_modules: set[ModuleType] = set()
+# modules that we've already seen and processed for local constants. These are
+# are all modules, not necessarily local ones. This lets us quickly see which
+# modules are new without an expensive path.resolve() or is_local_module_file
+# cache lookup.
+_seen_modules: set[ModuleType] = set()
+_sys_modules_len: Optional[int] = None
 
 
 def _get_local_constants() -> "ConstantsT":
+    global _sys_modules_len
+
+    if sys.platform == "emscripten":  # pragma: no cover
+        # pyodide builds bundle the stdlib in a nonstandard location, like
+        # `/lib/python312.zip/heapq.py`. To avoid identifying the entirety of
+        # the stdlib as local code and slowing down on emscripten, instead return
+        # that nothing is local.
+        #
+        # pyodide may provide some way to distinguish stdlib/third-party/local
+        # code. I haven't looked into it. If they do, we should correctly implement
+        # ModuleLocation for pyodide instead of this.
+        return _local_constants
+
     new_constants: set[ConstantT] = set()
-    new_modules = list(local_modules() - _local_modules)
-    for new_module in new_modules:
-        new_constants |= constants_from_module(new_module)
+    # We call this function once per HypothesisProvider instance, i.e. once per
+    # input, so it needs to be performant. The logic here is more complicated
+    # than necessary because of this.
+    #
+    # First, we check whether there are any new modules with a very cheap length
+    # check. This check can be fooled if a module is added while another module is
+    # removed, but the more correct check against tuple(sys.modules.keys()) is
+    # substantially more expensive. Such a new module would eventually be discovered
+    # if / when the length changes again in the future.
+    #
+    # If the length has changed, we find just modules we haven't seen before. Of
+    # those, we find the ones which correspond to local modules, and extract their
+    # constants.
+
+    # walrus operator avoids race conditions of sys.modules length changing from
+    # another thread before we set _sys_modules_len.
+    if (sys_modules_len := len(sys.modules)) != _sys_modules_len:
```

Comment:
I'm a bit confused by this comment. Is this any different from `sys_modules_len = len(sys.modules)` before the if statement?

(I don't object to the walrus operator, I'm just confused by the comment)